Flink 1.17: Support alter table column

linyanghao opened the pull request:
Resolves https://github.com/apache/iceberg/issues/7312

linyanghao commented:
For backward-compatibility, I'm currently remaining the old alterTable API as it is. Not sure if it's the right thing to do.
@hililiwei @stevenzwu @chenjunjiedada Could you take a look at this? Thanks!

chenjunjiedada commented on flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:
```diff
@@ -472,11 +479,13 @@ public void alterTable(ObjectPath tablePath, CatalogBaseTable newTable, boolean
     CatalogTable table = toCatalogTable(icebergTable);
 
-    // Currently, Flink SQL only support altering table properties.
+    // This alterTable API only supports altering table properties.
```
Should we deprecate this API?

linyanghao replied:
It seems that flink is not deprecating the old alterTable API yet. This is to support some legacy operations: `AlterTableSchemaOperation`, `AlterViewPropertiesOperation` and `AlterViewAsOperation`. 


Among these operations:

- `AlterTableSchemaOperation` is now [only used to alter hive tables](https://github.com/apache/flink/blob/553c72a894b25c10e71e54e6567c187876234fd5/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java#L133). 
- `AlterViewPropertiesOperation` and `AlterViewAsOperation` are for views.

So I guess it's safe to assume the old alterTable API won't be called by flink when altering iceberg tables.

linyanghao replied:
We cannot deprecate this API from iceberg's side, since it's defined in flink repo. But we can log some warnings inside this method instead. Is that ok?

chenjunjiedada replied:
That makes sense to me.

chenjunjiedada reviewed (approved):
+1. @stevenzwu you may also be interested in this.

linyanghao commented:
@chenjunjiedada Thanks for reviewing! I just fixed a bug on modifying the nullability of columns. You might want to check.

stevenzwu commented on flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/util/FlinkAlterTableUtil.java:
```diff
@@ -0,0 +1,176 @@
+        pendingUpdate.remove(resetOption.getKey());
+
+      } else {
+        throw new UnsupportedOperationException("Cannot apply unknown table change: " + change);
```
nit: error msg can be more specific to property changes.

stevenzwu commented on flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/util/FlinkAlterTableUtil.java:
```diff
@@ -0,0 +1,176 @@
+
+    } else {
+      throw new UnsupportedOperationException(
+          "Cannot apply unknown table change: " + modifyColumnPosition);
```
nit: more specific error msg

stevenzwu commented on flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/util/FlinkAlterTableUtil.java:
```diff
@@ -0,0 +1,176 @@
+          modifyComment.getOldColumn().getName(), modifyComment.getNewComment());
+
+    } else {
+      throw new UnsupportedOperationException("Cannot apply unknown table change: " + modifyColumn);
```
nit: more specific error msg

stevenzwu commented on flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:
```diff
@@ -472,11 +482,16 @@ public void alterTable(ObjectPath tablePath, CatalogBaseTable newTable, boolean
 
-    // For current Flink Catalog API, support for adding/removing/renaming columns cannot be done by
+    // Support for adding/removing/renaming columns cannot be done by
     // comparing
```
also this comment probably should be moved as method Javadoc

stevenzwu commented on flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:
```diff
@@ -517,6 +532,61 @@ public void alterTable(ObjectPath tablePath, CatalogBaseTable newTable, boolean
+        } else if ("current-snapshot-id".equalsIgnoreCase(set.getKey())) {
+          setSnapshotId = set.getValue();
+        } else if ("cherry-pick-snapshot-id".equalsIgnoreCase(set.getKey())) {
+          pickSnapshotId = set.getValue();
```
nit: `cherrypickSnapshotId` would be easier to understand

linyanghao replied:
This was copied from the old alterTable method. It was originally `pickSnapshotId`. But `cherrypickSnapshotId` is of course better. Let me fix this.

stevenzwu commented on flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:
```diff
@@ -605,6 +659,56 @@ private static void commitChanges(
+
+  private static void commitChanges(Table table, String setSnapshotId, String pickSnapshotId) {
+    // don't allow setting the snapshot and picking a commit at the same time because order is
+    // ambiguous and choosing
```
nit: fix line break

stevenzwu commented on flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:
```diff
@@ -605,6 +659,56 @@ private static void commitChanges(
+    transaction.commitTransaction();
+  }
+
+  private static void commitChanges(Table table, String setSnapshotId, String pickSnapshotId) {
```
this method name is a little misleading and confusing. maybe `applyManageSnapshots` would be more intuitive.

also curious why this method doesn't live inside `FlinkAlterTableUtil`?

linyanghao replied:
Maybe `commitManageSnapshots`?

stevenzwu commented on flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:
```diff
@@ -605,6 +659,56 @@ private static void commitChanges(
+    if (!schemaChanges.isEmpty()) {
+      UpdateSchema updateSchema = transaction.updateSchema();
+      FlinkAlterTableUtil.applySchemaChanges(updateSchema, schemaChanges);
+      updateSchema.commit();
```
maybe `commit` action can be moved inside `FlinkAlterTableUtil`? then it would be more natural to move `applyManageSnapshots` into `FlinkAlterTableUtil` too as mentioned in a comment below.

linyanghao replied:
That's reasonable.

hililiwei commented on flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/util/FlinkAlterTableUtil.java:
```diff
@@ -0,0 +1,176 @@
+        throw new UnsupportedOperationException("Modifying watermark specs is not supported yet. ");
+
+      } else if (change instanceof TableChange.DropWatermark) {
+        throw new UnsupportedOperationException("Watermark specs is not supported yet. ");
```
nit: the previous exception message indicated the action to be taken, such as add, modify, but there is no. Should unify it?

linyanghao replied:
Thanks for catching this!

hililiwei commented on flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTable.java:
```diff
@@ -357,6 +358,213 @@ public void testAlterTableWithPrimaryKey() throws TableNotExistException {
   }
 
+  @Test
+  public void testAlterTableProperties() throws TableNotExistException {
```
This method is very similar to testAlterTable(), they're doing the same thing, right?

linyanghao replied:
Yes, maybe I should remove the original testAlterTable()?

hililiwei replied:
yes, I think so.

hililiwei commented on flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTable.java:
```diff
@@ -357,6 +358,213 @@ public void testAlterTableWithPrimaryKey() throws TableNotExistException {
+  }
+
+  @Test
+  public void testAlterTableAddColumn() {
```
Do we need to add a case to test when the name of the field to be added already exists?

hililiwei commented on flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTable.java:
```diff
@@ -357,6 +358,213 @@ public void testAlterTableWithPrimaryKey() throws TableNotExistException {
+  }
+
+  @Test
+  public void testAlterTableDropColumn() {
```
Also, add a test to remove a nonexistent field?

linyanghao replied:
Sure, will add more tests.

linyanghao commented on flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:
```diff
@@ -472,11 +482,16 @@ public void alterTable(ObjectPath tablePath, CatalogBaseTable newTable, boolean
     // CatalogTable instances, unless the Flink schema contains Iceberg column IDs.
+
+    // To alter columns, use the other alterTable API and provide a list of TableChange's.
+    LOG.warn(
```
The user calling this method will pass a CatalogBaseTable argument, which represents how the 
user expect the table should be like after altering it. So I think it's better to warn the user, that the schema part of the CatalogBaseTable argument won't be effective.

linyanghao replied:
Sorry for missing out this comment earlier. I moved the message into the exception. You might want to check.

linyanghao commented:
@stevenzwu @hililiwei Thanks for reviewing! I added a few patches. Could you take a look?

linyanghao commented:
Hi @stevenzwu @hililiwei, do you have any more comments on this?

stevenzwu commented on flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTable.java:
```diff
@@ -322,39 +323,265 @@ public void testAlterTable() throws TableNotExistException {
+  }
+
+  @Test
+  public void testAlterTableModifyColumnPosition() {
```
can we add some failure tests? like non-exist column

linyanghao replied:
Added.

stevenzwu commented on flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTable.java:
```diff
@@ -322,39 +323,265 @@ public void testAlterTable() throws TableNotExistException {
+            .asStruct(),
+        schemaBefore.asStruct());
+
+    sql("ALTER TABLE tl MODIFY (dt STRING COMMENT 'some data')");
```
nit: comment as `comment for dt field`?

stevenzwu commented on flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTable.java:
```diff
@@ -322,39 +323,265 @@ public void testAlterTable() throws TableNotExistException {
+    // Setting an optional field as primary key should fail
+    Assertions.assertThatThrownBy(
+            () -> sql("ALTER TABLE tl MODIFY (PRIMARY KEY (col1) NOT ENFORCED)"))
+        .isInstanceOf(TableException.class);
```
nit: assert error msg too

stevenzwu commented on flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTable.java:
```diff
@@ -322,39 +323,265 @@ public void testAlterTable() throws TableNotExistException {
+    // Setting a composite key containing an optional field should fail
+    Assertions.assertThatThrownBy(
+            () -> sql("ALTER TABLE tl MODIFY (PRIMARY KEY (id, col1) NOT ENFORCED)"))
+        .isInstanceOf(TableException.class);
```
nit: assert error msg too

stevenzwu commented on flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTable.java:
```diff
@@ -322,39 +323,265 @@ public void testAlterTable() throws TableNotExistException {
+
+    // Dropping constraints is not supported yet
+    Assertions.assertThatThrownBy(() -> sql("ALTER TABLE tl DROP PRIMARY KEY"))
+        .isInstanceOf(TableException.class);
```
nit: assert error msg too

stevenzwu reviewed (commented):
@linyanghao this looks very close to me. just left some minor/nit comments

linyanghao commented on flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:
```diff
@@ -514,7 +534,68 @@ public void alterTable(ObjectPath tablePath, CatalogBaseTable newTable, boolean
+
+      } else if (change instanceof TableChange.ResetOption) {
+        propertyChanges.add(change);
+
```
Fixed. Sorry for missing this out earlier.

linyanghao commented on flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/util/FlinkAlterTableUtil.java:
```diff
@@ -0,0 +1,247 @@
+      Table table, String setSnapshotId, String cherrypickSnapshotId) {
+    // don't allow setting the snapshot and picking a commit at the same time because order is
+    // ambiguous and choosing
+    // one order leads to different results
```
I put "ambiguous and choosing" and "one order leads to different results" into one line. Did you mean this?

linyanghao commented:
@stevenzwu I addressed the comments. Regarding the warning message, I modified it so that users will only see it when they try altering schema using the old API. Could you take a look? Thanks!

stevenzwu commented on flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTable.java:
```diff
@@ -322,39 +323,293 @@ public void testAlterTable() throws TableNotExistException {
-        .get()
-        .alterTable(new ObjectPath(DATABASE, "tl"), catalogTable.copy(properties), false);
-    Assert.assertEquals(properties, table("tl").properties());
+    // Adding a required field should fail due to Iceberg's validation.
```
oh. this is because `SchemaUpdate` by default doesn't allow incompatible change. can you add the note to the comment for easier understanding for future reader.

linyanghao replied:
Added to all assertions on validation failures.

stevenzwu commented on flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/util/FlinkAlterTableUtil.java:
```diff
@@ -0,0 +1,176 @@
+        pendingUpdate.deleteColumn(dropColumn.getColumnName());
+
+      } else if (change instanceof TableChange.AddWatermark) {
+        throw new UnsupportedOperationException("Adding watermark specs is not supported yet. ");
```
this (the extra whitespace in the end) needs to be addressed.

stevenzwu replied:
also the error msg doesn't follow the coding style. you can referring to `ArrowVectorAccessor` for examples. Please review all the error msgs in this class

E.g., it could be sth like `Unsupported schema change: adding watermark specs`

stevenzwu commented:
@linyanghao I left a few comments. can you also rebase?

linyanghao commented:
@stevenzwu Hi, I rebased and addressed the style issues you mentioned. Could you check?

stevenzwu commented:
thanks @linyanghao for the contribution and @chenjunjiedada for the review.

@linyanghao can you create a back port PR? you can follow the example here for back port: https://github.com/apache/iceberg/pull/8228

linyanghao commented:
@stevenzwu Sure. Thanks for reviewing! Also thanks @chenjunjiedada and @hililiwei for reviewing!

linyanghao commented:
@stevenzwu Wait, Flink does not support ALTER TABLE in 1.16 yet. I guess we don't need to backport?

stevenzwu commented:
if that's the case, yeah no need to back port
